### PR TITLE
AMD Ryzen 7, Threadripper and EPIC CPU implementation.

### PR DESCRIPTION
**Pull request by aquacomputer**
CPU Temperatures, Voltages, Power, Clocks.

### CODE REVIEW

**Comment by Phyxion** [comment]
Great work 👍 

Few minor things though before accepting this PR:
- Can you remove the .editorconfig file.
- Can you fix some of the comments (most do not have a space before (and after) the //).
- Would it be possible to apply default C# naming convention (e.g., appendThread -> AppendThread, but also public should be capitalized (e.g., PackagePower) and privates should start with an underscore, e.g., _lastPwrTime. I know the OHM code base isn't very consistent with it but it'd be nice nonetheless. If you have ReSharper, it can do this automatically for you for the entire file.
- I don't think some of those fields/properties need to be public (e.g., PackagePower).  
- What's the reason for changing the thread affinity?

The rest looks good. I don't have a Ryzen myself unfortunately, so I can't test the result.

**Comment by aquacomputer** [comment]
> What's the reason for changing the thread affinity?
the zen virtualization feature need this.
everey thread can use the same cpu register but the result is specific for every core.

>Can you fix some of the comments
i can do it, but the base is a complete refactored OHM version: 
https://github.com/aquacomputer/openhardwaremonitor-aqua

**Comment by aquacomputer** [comment]
new version with code style changes ;)

File: Hardware/CPU/AMD17CPU.cs [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+        double energy = 15.3e-6 * pwr;
+        energy /= time.TotalSeconds;
+
+        Power.Value = (float)energy;
+        _hw.ActivateSensor(Power);
+
+      }
+      #endregion
+
+      public Sensor Clock { get; set; }
```
Comment (Phyxion) [inline Hardware/CPU/AMD17CPU.cs]: Comment above :)

File: Hardware/CPU/AMD17CPU.cs [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+          node.NodeId = numa_id;
+          node.Parent = this;
+          Nodes.Add(node);
+        }
+
+        if (thread != null)
+          node.AppendThread(thread, core_id);
+      }
+
+      public Sensor CoreTemperatureTctl { get; set; }
```
Comment (Phyxion) [inline Hardware/CPU/AMD17CPU.cs]: All sensors should be private (same with the sensors below). I don't think there's any hardware in OHM that has explicit sensors added like this.

**Comment by Phyxion** [comment]
Alright, makes sense indeed, thanks for improving. Left a few comments. 

I also have another question. Could the 'Processor' (mostly the same with Cores and Numa) be added to the AMD17CPU instead (which would be more in line with the other CPU classes)? It looks like its mostly grouping variables.

If you have any other useful features in the aqua base, we gladly accept changes. Would be nice if we could keep (most) changes/PRs in one place to avoid scattering. There's some changes I'd personally like to see in LHM, such as upgrade to .NET 4.5 and some code refactoring (at least apply a default style set in the project), but that would mean a more serious break from OHM. I'm not sure if @moellerm ever wants to merge some PRs back to OHM and doing so would make it much harder (which is the main reason I've not done this yet).

File: Hardware/CPU/AMD17CPU.cs [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+      {
+        this._hw = (AMD17CPU)hw;
+        Nodes = new List<NumaNode>();
+
+        PackagePower = new Sensor("Package Power", this._hw._sensorPower++, SensorType.Power, this._hw, this._hw.settings);
+        CoreTemperatureTctl = new Sensor("Core (Tctl)", this._hw._sensorTemperatures++, SensorType.Temperature, this._hw, this._hw.settings);
+        CoreTemperatureTdie = new Sensor("Core (Tdie)", this._hw._sensorTemperatures++, SensorType.Temperature, this._hw, this._hw.settings);
+        CoreVoltage = new Sensor("Core (SVI2)", this._hw._sensorVoltage++, SensorType.Voltage, this._hw, this._hw.settings);
+        SocVoltage = new Sensor("SoC (SVI2)", this._hw._sensorVoltage++, SensorType.Voltage, this._hw, this._hw.settings);
+        // node.coreCurrent = new Sensor("Core Current", hw.sensor_current++, SensorType.Current, hw, hw.settings);
```
[the remark this answers is not in the document]
Comment (sebastian-dev) [reply Hardware/CPU/AMD17CPU.cs]: The CPU has an current sensor, but it is not cler how to read the sensor

File: Hardware/CPU/AMD17CPU.cs [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+using System.IO;
+using System.Text;
+using System.Threading;
+
+namespace OpenHardwareMonitor.Hardware.CPU
+{
+  internal sealed class AMD17CPU : AMDCPU
+  {
+    // counter, to create sensor index values 
+    public int _sensorTemperatures = 0;
```
[the remark this answers is not in the document]
Comment (sebastian-dev) [reply Hardware/CPU/AMD17CPU.cs]: ok, all sensors can change to private

File: Hardware/CPU/AMD17CPU.cs [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+        CoreVoltage = new Sensor("Core (SVI2)", this._hw._sensorVoltage++, SensorType.Voltage, this._hw, this._hw.settings);
+        SocVoltage = new Sensor("SoC (SVI2)", this._hw._sensorVoltage++, SensorType.Voltage, this._hw, this._hw.settings);
+        // node.coreCurrent = new Sensor("Core Current", hw.sensor_current++, SensorType.Current, hw, hw.settings); 
+        // node.socCurrent = new Sensor("SoC Current", hw.sensor_current++, SensorType.Current, hw,hw.settings); 
+      }
+
+      #region UpdateSensors
+      public void UpdateSensors()
+      {
+        // var node = nodes.FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (sebastian-dev) [reply Hardware/CPU/AMD17CPU.cs]: this is the back port to net 2.0, in my net 4.6 version i use linq expressions

**Comment by sebastian-dev** [comment]
every processor can have up to 4 numa nodes with 8 codes (16 threads) when i read out the cpu i have to know the node (numa) and the thread routing (thread-> node -> core id).
I think this nesting simplify upcomming different cpu variants.
2 numa cores with 2x6 (12 core) or any other cpu configuration.

**Comment by aquacomputer** [comment]
all changes are online, i removed not used values

**Comment by Phyxion** [comment]
Will look at it later today, thank you very much @aquacomputer.

File: Hardware/CPU/AMD17CPU.cs [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    private int _sensorMulti = 0;
+
+    // register index names for CPUID[] 
+    private const int EAX = 0;
+    private const int EBX = 1;
+    private const int ECX = 2;
+    private const int EDX = 3;
+
+    #region amd zen registers 
+    public const uint PERF_CTL_0 = 0xC0010000;
```
Comment (Phyxion) [inline Hardware/CPU/AMD17CPU.cs]: All these registers can be private (same as IntelCPU).

File: Hardware/CPU/AMD17CPU.cs [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+
+        // update for next sample 
+        _lastPwrTime = sample_time;
+        _lastPwrValue = total_energy;
+
+        double energy = 15.3e-6 * pwr;
+        energy /= time.TotalSeconds;
+
+        _packagePower.Value = (float)energy;
+        _hw.ActivateSensor(_packagePower);
```
Comment (Phyxion) [inline Hardware/CPU/AMD17CPU.cs]: While OHM's ListSet (underlying collection for Sensors) doesn't add duplicate sensors, I think this ActivateSensor can be moved to the constructor as it always exists (same with the other 5 sensors in the constructor).

**Comment by Phyxion** [comment]
Awesome, left some very minor things and then I'll merge it in. 👍 I can also do those changes if you want (not sure if that's easier with your own base).

**Comment by sebastian-dev** [comment]
Changes committed.

Why do you not use a .editorconfig file. The most Visual Studio configurations have other code styles as default.

**Comment by Phyxion** [comment]
Awesome, thanks @sebastian-dev.
Mostly because the code style it's not very consistent as-is, and the some fixing before it is applied. I'm giving @moellerm just a little bit more time to see if he wants to merge anything in (and add some collaborators). If nothing by the end of the year I'll put ReSharper over the project to clean everything up at once.

**Pull request merged**